cirq/work/observable_settings.py

```diff
@@ -0,0 +1,186 @@
+    """
+    qubit_pauli_map = dict()  # type: Dict[ops.Qid, ops.Pauli]
+    for observable in observables:
+        for qubit, pauli in observable.items():
+            if qubit in qubit_pauli_map:
+                if qubit_pauli_map[qubit] != pauli:
+                    return None
+            else:
+                qubit_pauli_map[qubit] = pauli
+    return ops.PauliString(qubit_pauli_map)
+
+
+def _max_weight_state(states: Iterable[value.ProductState]) \
+        -> Union[None, value.ProductState]:
+    """Create a new state that is compatible with all input states
+    and has the maximum weight.
+
+    The returned TensorProductState is constructed by taking the
+    single-qubit state at each qubit position.
+
+    This function will return `None` if the input states are not compatible
+
+    For example, the max_weight_state of [+X(0), -Z(1)] is
+    "+X(0) * -Z(1)". Asking for the max weight state of something like
+    [+X(0), +Z(0)] will return None.
+    """
+    qubit_state_map = dict()  # type: Dict[ops.Qid, _NamedOneQubitState]
+    for state in states:
+        for qubit, named_state in state:
+            if qubit in qubit_state_map:
+                if qubit_state_map[qubit] != named_state:
+                    return None
+            else:
+                qubit_state_map[qubit] = named_state
+    return value.ProductState(qubit_state_map)
+
+
+def zeros_state(qubits: Iterable['cirq.Qid']):
+    """Return the ProductState that is |00..00> on all qubits."""
+    return value.ProductState({q: value.KET_ZERO for q in qubits})
+
+
+def observables_to_settings(observables: Iterable['cirq.PauliString'],
+                            qubits: Iterable['cirq.Qid']
+                           ) -> Iterable[InitObsSetting]:
+    """Transform an observable to an InitObsSetting initialized in the
+    all-zeros state.
+    """
+    for observable in observables:
+        yield InitObsSetting(init_state=zeros_state(qubits),
+                             observable=observable)
+
+
+def _fix_precision(val: float, precision) -> int:
+    """Convert floating point numbers to (implicitly) fixed point integers.
+
+    Circuit parameters can be floats but we also need to use them as
+    dictionary keys. We secretly use these fixed-precision integers.
+    """
+    return int(val * precision)
+
+
+def _hashable_param(param_tuples: Iterable[Tuple[str, float]], precision=1e7):
+    """Hash circuit parameters using fixed precision.
+
+    Circuit parameters can be floats but we also need to use them as
+    dictionary keys. We secretly use these fixed-precision integers.
+    """
+    return frozenset((k, _fix_precision(v, precision)) for k, v in param_tuples)
+
+
+@protocols.json_serializable_dataclass(frozen=True)
+class MeasurementSpec:
+    """An encapsulation of all the specifications for one run of a
+    quantum processor.
+
+    This includes the maximal input-output setting (which may result in many
```

Comment:
I think we need to provide more context, because we are introducing a top level class. 

Especially that it will come up right next to `cirq.MeasurementGate` - it would be interesting to explain the context of observable measurements. 

![image](https://user-images.githubusercontent.com/635073/95514226-95016c00-0989-11eb-9b80-31fed1e106fa.png)

We should also explain a bit more what `max_setting` and `circuit_params` are.

Reply:
Yeah, I think I would prefer MeasurementSpec to be private. I'm not 100% sure how that would work with Cirq's `repr` tests and json serialization.

Reply:
made it private